src/phoenix/utilities/ratelimits.py

```diff
@@ -0,0 +1,210 @@
+import asyncio
+import sys
+import time
+from collections import defaultdict
+from functools import wraps
+from typing import Any, Awaitable, Callable, Dict, Optional, TypeVar, Union, cast
+
+if sys.version_info < (3, 10):
+    from typing_extensions import ParamSpec
+else:
+    from typing import ParamSpec
+
+Numeric = Union[int, float]
+P = ParamSpec("P")
+T = TypeVar("T")
+A = TypeVar("A", bound=Callable[..., Awaitable[Any]])
+
+
+class UnavailableTokensError(Exception):
+    pass
+
+
+class TokenLimiter:
+    """
+    A simple in-memory rate-limiter implemented using the leaky bucket algorithm.
+    """
+
+    def __init__(
+        self,
+        per_minute_rate: Numeric,
+        starting_tokens: Numeric,
+        max_tokens: Numeric,
+        rate_multiplier: Numeric = 1,
+    ):
+        self.rate_multiplier = rate_multiplier
+        self.rate = self._per_second_rate(per_minute_rate)
+        self.tokens = starting_tokens
+        self.max_tokens = max_tokens
+        self.created = time.time()
+        self.last_checked = self.created
+        self.total_tokens: Numeric = 0
+
+    def _per_second_rate(self, per_minute_rate: Numeric) -> float:
+        return round(per_minute_rate / 60, 3) * self.rate_multiplier
+
+    def refresh_limit(self, per_minute_rate: Numeric, max_tokens: Numeric) -> None:
+        new_rate = self._per_second_rate(per_minute_rate)
```

Comment:
why does the per-second rate need to be set?

Reply:
`token_rate_limiter.rate` is in units of `tokens/sec`, we just do the conversion up front, but your comments are making me think that it's probably better to just push the responsibility of conversion out to the caller to avoid unit weirdness at all